### PR TITLE
Introducing solution file for test projects

### LINKED ISSUE #5099: Add test sln for CLI
CLI currently restores each test indepenently, instead of doing a sln-restore that can share context from project to project. We should add a .sln file in the test/ directory that references all of our test projects, then change build scripts to restore that single file instead of restoring each test project.

### PR DESCRIPTION
**Pull request by blackdwarf**
This commit introduces a solution file for test projects and modifies
the needed build targets to restore the test solution instead of each
test project separately.

It also introduces a new template called "sln" to dotnet new that drops
a template solution file for users to consume.

Fixes #5099

### CODE REVIEW

File: src/dotnet/dotnet.csproj [hunk 1]
```diff
@@ -13,9 +13,9 @@
...
     <EmbeddedResource Include="**\*.resx" Exclude="bin\**;obj\**;**\*.xproj;packages\**" />
-    <EmbeddedResource Include="commands\dotnet-new\CSharp_Console.zip;commands\dotnet-new\CSharp_Lib.zip;commands\dotnet-new\CSharp_Mstest.zip;commands\dotnet-new\CSharp_Xunittest.zip;commands\dotnet-new\CSharp_Web.zip;compiler\resources\**\*" Exclude="bin\**;obj\**;**\*.xproj;packages\**" />
+    <EmbeddedResource Include="commands\dotnet-new\CSharp_Sln.zip;commands\dotnet-new\CSharp_Console.zip;commands\dotnet-new\CSharp_Lib.zip;commands\dotnet-new\CSharp_Mstest.zip;commands\dotnet-new\CSharp_Xunittest.zip;commands\dotnet-new\CSharp_Web.zip;compiler\resources\**\*" Exclude="bin\**;obj\**;**\*.xproj;packages\**" />
     <EmbeddedResource Include="commands\dotnet-new\FSharp_Console.zip;commands\dotnet-new\FSharp_Lib.zip;commands\dotnet-new\FSharp_Mstest.zip;commands\dotnet-new\FSharp_Xunittest.zip;commands\dotnet-new\FSharp_Web.zip" Exclude="bin\**;obj\**;**\*.xproj;packages\**" />
```
Comment (jgoshi) [inline src/dotnet/dotnet.csproj]: Do you also need to include FSharp_Sln.zip?

File: src/dotnet/commands/dotnet-new/FSharp_Sln/$projectName$.sln [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Microsoft Visual Studio Solution File, Format Version 15.00
```
Comment (livarcocc) [inline src/dotnet/commands/dotnet-new/FSharp_Sln/$projectName$.sln]: Shouldn't this be a language agnostic template?

File: src/dotnet/commands/dotnet-new/Program.cs [hunk 3]
```diff
@@ -140,7 +140,8 @@ public static int Run(string[] args)
                                    new { Name = "Web", isMsBuild = true }, 
                                    new { Name = "Lib", isMsBuild = true },
                                    new { Name = "Mstest", isMsBuild = true },
-                                   new { Name = "Xunittest", isMsBuild = true }
+                                   new { Name = "Xunittest", isMsBuild = true },
```
Comment (livarcocc) [inline src/dotnet/commands/dotnet-new/Program.cs]: Just to confirm, there was nothing that was needed to allow this template to run on a non-empty folder?

File: src/dotnet/dotnet.csproj (hunk 1 above)
Comment (blackdwarf) [inline src/dotnet/dotnet.csproj]: Yep, good catch.

File: src/dotnet/commands/dotnet-new/FSharp_Sln/$projectName$.sln (hunk 2 above)
Comment (blackdwarf) [inline src/dotnet/commands/dotnet-new/FSharp_Sln/$projectName$.sln]: We don't have support for that currently in `dotnet new` so I just used the existing infrastructure. I can endevor to make this support, if you think it is OK. :)

File: src/dotnet/commands/dotnet-new/Program.cs (hunk 3 above)
Comment (blackdwarf) [inline src/dotnet/commands/dotnet-new/Program.cs]: @livarcocc tested it out, and nope. The logic is such that it checks if the file with the same name (e.g. `Program.cs`) exists in the folder and errors out on that.

**Comment by livarcocc** [comment]
@dotnet-bot Test CentOS7.1 x64 Debug Build

**Comment by blackdwarf** [comment]
@livarcocc for sure, I need to figure out the `TestAssets` thing and will add a couple of tests.

File: src/dotnet/commands/dotnet-new/CSharp_Sln/$projectName$.sln [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+# Visual Studio 15
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|Any CPU = Debug|Any CPU
+		Release|Any CPU = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+EndGlobal
```
Comment (piotrpMSFT) [inline src/dotnet/commands/dotnet-new/CSharp_Sln/$projectName$.sln]: please add a trailing newline.

File: src/dotnet/commands/dotnet-new/FSharp_Sln/$projectName$.sln [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+# Visual Studio 15
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|Any CPU = Debug|Any CPU
+		Release|Any CPU = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+EndGlobal
```
Comment (piotrpMSFT) [inline src/dotnet/commands/dotnet-new/FSharp_Sln/$projectName$.sln]: Here as well.

**Comment by blackdwarf** [comment]
@dotnet-bot test this please

File: build/Microsoft.DotNet.Cli.Test.targets [hunk 6]
```diff
@@ -90,7 +90,6 @@
           Inputs="@(RestoreTestsInputs)"
           Outputs="@(RestoreTestsInputs->'%(RelativeDir)/obj/project.assets.json');@(RestoreTestsInputs->'%(RelativeDir)/obj/%(Filename).csproj.nuget.g.props');@(RestoreTestsInputs->'%(RelativeDir)/obj/%(Filename).csproj.nuget.g.targets')">
     <CallTarget Targets="CleanTestProjectsLockFiles"/>
-
```
Comment (piotrpMSFT) [inline build/Microsoft.DotNet.Cli.Test.targets]: ?

File: build/Microsoft.DotNet.Cli.Test.targets [hunk 7]
```diff
@@ -112,7 +111,7 @@
   <Target Name="SetupRestoreTestsInputs"
           DependsOnTargets="PrepareTests;" >
     <ItemGroup>
-      <RestoreTestsInputs Include="$(TestDirectory)/**/*.csproj"
+      <RestoreTestsInputs Include="$(TestDirectory)/Microsoft.DotNet.Cli.Tests.sln"
```
Comment (piotrpMSFT) [inline build/Microsoft.DotNet.Cli.Test.targets]: You can remove the exclude. Also, I don't think we need the itemgroup anymore and can just pass this path to <DotNetRestore> as is, no?

File: build/Microsoft.DotNet.Cli.Test.targets (hunk 6 above)
Comment (blackdwarf) [inline build/Microsoft.DotNet.Cli.Test.targets]: Have no idea what this is...

File: build/Microsoft.DotNet.Cli.Test.targets (hunk 7 above)
Comment (blackdwarf) [inline build/Microsoft.DotNet.Cli.Test.targets]: Well, I wanted to do as little as possible to disrupt the current state. Lemme try without it and see where we end up...

**Comment by blackdwarf** [comment]
@dotnet-bot test this please

**Comment by blackdwarf** [comment]
@livarcocc is it ok to check this in without a test since I removed the new template for an empty SLN?

**Comment by blackdwarf** [comment]
@piotrpMSFT you OK with the change?

**Comment by blackdwarf** [comment]
@dotnet-bot test Windows_NT x86 Debug Build please

File: build/Microsoft.DotNet.Cli.Test.targets (hunk 6 above)
Comment (piotrpMSFT) [inline build/Microsoft.DotNet.Cli.Test.targets]: Can you put back the newline that was removed?

**Comment by piotrpMSFT** [comment]
@MattGertz I'm merging this infra-only change. It takes advantage of NuGet's ability to restore an entire solution to reduce both code complexity and runtime of test project restore in CLI.

**Pull request merged**